File: java/P0763.java
/**
 * P0763:划分字母区间
 * 题意：字符串 S 由小写字母组成。我们要把这个字符串划分为尽可能多的片段，同一个字母只会出现在其中的一个片段。返回一个表示每个字符串片段的长度的列表。
 * 题解：
 */
class P0763{
    public static void main(String[] args) {
        
    }
    static class Solution{
        public List<Integer> partitionLabels(String S) {
            List<Integer> ans = new ArrayList<>();
            int[] lastIndex = new int[26];
            int strLen = S.length();
            for(int i =0;i<strLen;i++){
                lastIndex[(int)(S.charAt(i) - 'a')] = i;
            }
            int now = 0,i = 0;
            while(now < strLen){
                int nowlen = 0;
                int j = lastIndex[(int)(S.charAt(i) - 'a')];
                while(i <= j){
                    j = Math.max(j,lastIndex[(int)(S.charAt(i) - 'a')]);
                    nowlen++;
                    i++;
                    now++;
                }
                ans.add(nowlen);
            }
           return ans;
        }
    }
}

File: go/P0763.go
//题意：划分字符串片段，让相同的字符出现在同一片段
//题解：两次for循环
package main

import "fmt"

func partitionLabels(S string) []int {
    ans := make([]int, 0)
    lastIndex := make([]int, 26)
    for i,v := range(S){
        lastIndex[int(v-'a')] = i
    }
    fmt.Println(lastIndex)
    beg, end := -1,0
    for i,v := range(S){
        if end > lastIndex[int(v-'a')]{
            end = end
        }else{
            end = lastIndex[int(v-'a')]
        }
        fmt.Println(beg,end)
        if end == i{
            ans = append(ans, end-beg)
            beg = end
        }
    }
    return ans
}